src/libraries/System.Private.CoreLib/src/System/Threading/Lock.cs

```diff
@@ -0,0 +1,1194 @@
+                }
+            }
+
+            private bool HasAnySpinners => (_state & SpinnerCountMask) != 0;
+
+            private bool TryIncrementSpinnerCount()
+            {
+                uint newState = _state + SpinnerCountIncrement;
+                if (new State(newState).HasAnySpinners) // overflow check
+                {
+                    _state = newState;
+                    return true;
+                }
+                return false;
+            }
+
+            private void DecrementSpinnerCount()
+            {
+                Debug.Assert(HasAnySpinners);
+                _state -= SpinnerCountIncrement;
+            }
+
+            private bool IsWaiterSignaledToWake => (_state & IsWaiterSignaledToWakeMask) != 0;
+
+            private void SetIsWaiterSignaledToWake()
+            {
+                Debug.Assert(HasAnyWaiters);
+                Debug.Assert(NeedToSignalWaiter);
+
+                _state += IsWaiterSignaledToWakeMask;
+            }
+
+            private void ClearIsWaiterSignaledToWake()
+            {
+                Debug.Assert(IsWaiterSignaledToWake);
+                _state -= IsWaiterSignaledToWakeMask;
+            }
+
+            public bool HasAnyWaiters => _state >= WaiterCountIncrement;
+
+            private bool TryIncrementWaiterCount()
+            {
+                uint newState = _state + WaiterCountIncrement;
+                if (new State(newState).HasAnyWaiters) // overflow check
+                {
+                    _state = newState;
+                    return true;
+                }
+                return false;
+            }
+
+            private void DecrementWaiterCount()
+            {
+                Debug.Assert(HasAnyWaiters);
+                _state -= WaiterCountIncrement;
+            }
+
+            public bool NeedToSignalWaiter
+            {
+                get
+                {
+                    Debug.Assert(HasAnyWaiters);
+                    return (_state & (SpinnerCountMask | IsWaiterSignaledToWakeMask)) == 0;
+                }
+            }
+
+            public static bool operator ==(State state1, State state2) => state1._state == state2._state;
+            public static bool operator !=(State state1, State state2) => !(state1 == state2);
+
+            bool IEquatable<State>.Equals(State other) => this == other;
+            public override bool Equals(object? obj) => obj is State other && this == other;
+            public override int GetHashCode() => (int)_state;
+
+            private static State CompareExchange(Lock lockObj, State toState, State fromState) =>
+                new State(Interlocked.CompareExchange(ref lockObj._state, toState._state, fromState._state));
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static bool TryLock(Lock lockObj)
+            {
+                // The lock is fair to release waiters in FIFO order, but allows non-waiters to acquire the lock if it's
```

Comment:
Is this comment actually correct? Perhaps only on Unix, where we serialize the threads.

On Windows most synchronization primitives do not guarantee any order. I think some fairness is expected - like every blocked thread will get to run, eventually, but the docs explicitly say it is not FIFO.

I think what we can guarantee is just that no waiters will be infinitely starved while other threads keep acquiring/releasing the lock.

Reply:
Auto-reset event on Windows is mostly FIFO, which can be seen from a simple test, but as described here the wait  order can change:

https://learn.microsoft.com/en-us/windows/win32/sync/event-objects
> "If more than one thread is waiting, a waiting thread is selected. Do not assume a first-in, first-out (FIFO) order. External events such as kernel-mode APCs can change the wait order."

APCs would change the order, and a thread that is awakened but does not acquire the lock would wait again but goes to the back of the queue, which is also not fully fair.

It's not guaranteed FIFO, but it's being used because it's FIFO-enough that it's reasonable to use for this purpose. I'll update the wording, I don't intend on guaranteeing anything, only to indicate what would be typical.